### ISSUE TITLE
Paste slow in WSL

### ISSUE THREAD

**Issue by rickywu** [neovim/neovim#21739]
### Describe the bug

set clipboard=unnamedplus

then press more than 3 times p, get paste very slow compare with set clipboard=""

### Steps to reproduce

nvim --clean
:edit foo
:set clipboard=unnamedplus
input something and yank
press more than 3 times p to paste

### Expected behavior

paste run very slow compare with set clipboard=""
more times paste get more obviously

### Neovim version (nvim -v)

0.9.0 and 0.8.2

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Windows 11 WSL2

### Terminal name/version

mobaxterm

### $TERM environment variable

xterm-256

### Installation

apt install nvim-linux.deb

**Comment by 3N4N** [comment]
What's the output of `:set g:clipboard?`?  Could you try setting it to `:h clipboard-wsl`?  If setting it to ':h clipboard-wsl' still makes pasting slow, try this (make sure 'win32yank.exe' is in $PATH or use the absolute path):

```
  let g:clipboard = {
        \   'name': 'win32yank_nvim',
        \   'copy': {
        \      '+': 'win32yank.exe -i --crlf',
        \      '*': 'win32yank.exe -i --crlf',
        \    },
        \   'paste': {
        \      '+': 'win32yank.exe -o --lf',
        \      '*': 'win32yank.exe -o --lf',
        \   },
        \   'cache_enabled': 1,
        \ }
```

A reason for pasting to be slow, even with proper 'g:clipboard', is maybe invoking powershell takes a long time.  If that's the case, using 'win32yank.exe' should make it significantly faster.

**Comment by rickywu** [comment]
Actually I used this config before verify this problem
![image](https://user-images.githubusercontent.com/442398/211736120-30c74cef-8a59-47e9-8953-ab7d75bc27f1.png)

Them same, and set clipboard="" can solve

**Comment by 3N4N** [comment]
> set clipboard="" can solve

That's because with `:set clipboard=` you're using vim's internal registers, without resorting to external executables.

> now I can use <C-C> and <C-V> to copy and paste between WLS and Windows

I'm not sure what <C-C> and <C-V> means.  Are you referring to the copy-paste facility of your terminal emulator?  If so, then you're bypassing Vim entirely.  Not a bad idea.

**Comment by rickywu** [comment]
This is not relate to external executables, obviously is a performance problem of nvim

**Comment by justinmk** [comment]
> set clipboard=unnamedplus
> 
> then press more than 3 times p, get paste very slow

sounds different than https://github.com/neovim/neovim/issues/11804 , which is about single-operation performance. If "p" is slow after the 3rd try, it sounds like state is being held somewhere, or requests are queuing up...

**Comment by rickywu** [comment]
3 times is not the problem, only once still can feel slow than normal, I mean 3 times can give you an obviously result

**Comment by EtiamNullam** [comment]
It doesn't apply only to `WSL`. Surely every `Windows` (and `win32yank`) user suffers this. As I've mentioned in the other issue: I experience 60-100ms delay on paste (put) with `clipboard=unnamed*`, so I guess that's roughly the overhead for accessing the system's clipboard (which I believe has to be done on every `put`) with `win32yank`.

Same for setting `clipboard=unnamed*` on startup - it will increase the startup time of `neovim` by that 60-100ms.

NOTE: by `clipboard=unnamed*` I mean options `clipboard=unnamed` or `clipboard=unnamedplus`.

I feel like it's best to delay any clipboard synchronization until `neovim` is out of focus, as you've mentioned @justinmk, sadly not every terminal and UI supports focus change events, like I did in the plugin that I've created: https://github.com/EtiamNullam/deferred-clipboard.nvim

I'm happy that there is another issue bringing up the problem to the front, as the previous was locked down which surely slowed down any progress on it to a crawl.

**Comment by justinmk** [comment]
> only once still can feel slow than normal

then this is a duplicate of https://github.com/neovim/neovim/issues/11804 . Please search existing issues before creating new ones.



> I'm happy that there is another issue bringing up the problem to the front, as the previous was locked down which surely slowed down any progress on it to a crawl.

 https://github.com/neovim/neovim/issues/11804  is locked because new comments aren't providing info. What's needed is a PR. 

We didn't get new information here.

**Comment by yioneko** [comment]
For anyone only want a workaround for win32yank, here is my solution for this. Unlike [deferred-clipboard.nvim](https://github.com/EtiamNullam/deferred-clipboard.nvim), this uses job control to synchronize clipboard asynchronously and doesn't rely on focus event for copy (but require it on paste), thus never block editor. 
```lua
local M = {}

local augroup = A.nvim_create_augroup("deferClip", {})

local entries = {
  first = 1,
  last = 1,
}
local active_entry = {}

local function add_entry(entry)
  entries[entries.last] = entry
  entries.last = entries.last + 1
end

local function pop_entry()
  if entries.first < entries.last then
    local entry = entries[entries.first]
    entries[entries.first] = nil
    entries.first = entries.first + 1
    return entry
  end
end

local function sync_from()
  vim.fn.jobstart({ "win32yank.exe", "-o", "--lf" }, {
    stdout_buffered = true,
    on_stdout = function(_, data)
      active_entry = { lines = data, regtype = "v" }
    end,
  })
end

local sync_to
do
  local cur_sync_job
  local function sync_next(entry)
    local chan = vim.fn.jobstart({ "win32yank.exe", "-i" }, {
      on_exit = function(_)
        local next_entry = pop_entry()
        if next_entry then
          sync_next(next_entry)
        else
          cur_sync_job = nil
        end
      end,
    })
    cur_sync_job = chan
    vim.fn.chansend(chan, entry.lines)
    vim.fn.chanclose(chan, "stdin")
  end

  sync_to = function()
    if cur_sync_job then
      return
    else
      local entry = pop_entry()
      if entry then
        sync_next(entry)
      end
    end
  end
end

function M.copy(lines, regtype)
  add_entry({ lines = lines, regtype = regtype })
  active_entry = { lines = lines, regtype = regtype }
  sync_to()
end

function M.get_active()
  return { active_entry.lines, active_entry.regtype }
end

function M.setup()
  vim.api.nvim_exec(
    [[
    function s:copy(lines, regtype)
      call luaeval('require("core.vim.deferclip").copy(_A[1], _A[2])', [a:lines, a:regtype])
    endfunction
    function s:get_active()
      call luaeval('require("core.vim.deferclip").get_active()')
    endfunction

    let g:clipboard = {
          \   'name': 'deferClip',
          \   'copy': {
          \      '+': {lines, regtype -> s:copy(lines, regtype)},
          \      '*': {lines, regtype -> s:copy(lines, regtype)},
          \    },
          \   'paste': {
          \      '+': {-> s:get_active()},
          \      '*': {-> s:get_active()},
          \   },
          \ }
  ]],
    false
  )
  A.nvim_create_autocmd({ "FocusGained", "VimEnter" }, {
    group = augroup,
    callback = sync_from,
  })
end

return M
```

**Comment by justinmk** [comment]
@yioneko that would be a welcome PR to Nvim core. That is what we need for https://github.com/neovim/neovim/issues/11804

**Comment by BartSte** [comment]
> For anyone only want a workaround for win32yank, here is my solution for this. Unlike [deferred-clipboard.nvim](https://github.com/EtiamNullam/deferred-clipboard.nvim), this uses job control to synchronize clipboard asynchronously and doesn't rely on focus event for copy (but require it on paste), thus never block editor.
> 
> ```lua
> local M = {}
> 
> local augroup = A.nvim_create_augroup("deferClip", {})
> 
> local entries = {
>   first = 1,
>   last = 1,
> }
> local active_entry = {}
> 
> local function add_entry(entry)
>   entries[entries.last] = entry
>   entries.last = entries.last + 1
> end
> 
> local function pop_entry()
>   if entries.first < entries.last then
>     local entry = entries[entries.first]
>     entries[entries.first] = nil
>     entries.first = entries.first + 1
>     return entry
>   end
> end
> 
> local function sync_from()
>   vim.fn.jobstart({ "win32yank.exe", "-o", "--lf" }, {
>     stdout_buffered = true,
>     on_stdout = function(_, data)
>       active_entry = { lines = data, regtype = "v" }
>     end,
>   })
> end
> 
> local sync_to
> do
>   local cur_sync_job
>   local function sync_next(entry)
>     local chan = vim.fn.jobstart({ "win32yank.exe", "-i" }, {
>       on_exit = function(_)
>         local next_entry = pop_entry()
>         if next_entry then
>           sync_next(next_entry)
>         else
>           cur_sync_job = nil
>         end
>       end,
>     })
>     cur_sync_job = chan
>     vim.fn.chansend(chan, entry.lines)
>     vim.fn.chanclose(chan, "stdin")
>   end
> 
>   sync_to = function()
>     if cur_sync_job then
>       return
>     else
>       local entry = pop_entry()
>       if entry then
>         sync_next(entry)
>       end
>     end
>   end
> end
> 
> function M.copy(lines, regtype)
>   add_entry({ lines = lines, regtype = regtype })
>   active_entry = { lines = lines, regtype = regtype }
>   sync_to()
> end
> 
> function M.get_active()
>   return { active_entry.lines, active_entry.regtype }
> end
> 
> function M.setup()
>   vim.api.nvim_exec(
>     [[
>     function s:copy(lines, regtype)
>       call luaeval('require("core.vim.deferclip").copy(_A[1], _A[2])', [a:lines, a:regtype])
>     endfunction
>     function s:get_active()
>       call luaeval('require("core.vim.deferclip").get_active()')
>     endfunction
> 
>     let g:clipboard = {
>           \   'name': 'deferClip',
>           \   'copy': {
>           \      '+': {lines, regtype -> s:copy(lines, regtype)},
>           \      '*': {lines, regtype -> s:copy(lines, regtype)},
>           \    },
>           \   'paste': {
>           \      '+': {-> s:get_active()},
>           \      '*': {-> s:get_active()},
>           \   },
>           \ }
>   ]],
>     false
>   )
>   A.nvim_create_autocmd({ "FocusGained", "VimEnter" }, {
>     group = augroup,
>     callback = sync_from,
>   })
> end
> 
> return M
> ```
Hey Yioneko,

If found that the table `A` is not defined when I tried running your script. Can you point me to thing I am missing here?

**Comment by yioneko** [comment]
@BartSte It's `vim.api`. I've corrected it.

**Comment by BartSte** [comment]
Thanks!
